diff -ur plug-ins/bmp.orig/bmp.c plug-ins/bmp/bmp.c
--- plug-ins/bmp.orig/bmp.c	2005-12-30 11:44:08.000000000 -0500
+++ plug-ins/bmp/bmp.c	2005-12-30 12:43:05.000000000 -0500
@@ -109,6 +109,7 @@
     { GIMP_PDB_DRAWABLE, "drawable",     "Drawable to save" },
     { GIMP_PDB_STRING,   "filename",     "The name of the file to save the image in" },
     { GIMP_PDB_STRING,   "raw_filename", "The name entered" },
+    { GIMP_PDB_INT32,    "rle_encoded",  "0 for uncompressed, 1 for RLE compression (only used for Indexed images)" },
   };
 
   gimp_install_procedure ("file_bmp_load",
@@ -238,7 +239,7 @@
         case GIMP_RUN_NONINTERACTIVE:
           /*  Make sure all the arguments are there!  */
           interactive_bmp = FALSE;
-	  if (nparams != 5)
+	  if (nparams != 6)
             status = GIMP_PDB_CALLING_ERROR;
           break;
 
@@ -252,7 +253,7 @@
 
       if (status == GIMP_PDB_SUCCESS)
 	{
-	  status = WriteBMP (param[3].data.d_string, image_ID, drawable_ID);
+	  status = WriteBMP (param[3].data.d_string, image_ID, drawable_ID, param[5].data.d_int32);
 	}
 
       if (export == GIMP_EXPORT_EXPORT)
diff -ur plug-ins/bmp.orig/bmp.h plug-ins/bmp/bmp.h
--- plug-ins/bmp.orig/bmp.h	2005-12-30 11:44:08.000000000 -0500
+++ plug-ins/bmp/bmp.h	2005-12-30 11:55:47.000000000 -0500
@@ -28,7 +28,8 @@
 gint32             ReadBMP   (const gchar *filename);
 GimpPDBStatusType  WriteBMP  (const gchar *filename,
                               gint32       image,
-                              gint32       drawable_ID);
+                              gint32       drawable_ID,
+                              gint32       rle_encoded);
 
 extern       gboolean  interactive_bmp;
 extern const gchar    *filename;
diff -ur plug-ins/bmp.orig/bmpwrite.c plug-ins/bmp/bmpwrite.c
--- plug-ins/bmp.orig/bmpwrite.c	2005-12-30 11:44:08.000000000 -0500
+++ plug-ins/bmp/bmpwrite.c	2005-12-30 12:05:03.000000000 -0500
@@ -105,7 +105,8 @@
 GimpPDBStatusType
 WriteBMP (const gchar *filename,
           gint32       image,
-          gint32       drawable_ID)
+          gint32       drawable_ID,
+          gint32       rle_encoded)
 {
   FILE         *outfile;
   gint          Red[MAXCOLORS];
@@ -185,10 +186,17 @@
 
   /* Perhaps someone wants RLE encoded Bitmaps */
   encoded = 0;
-  if ((BitsPerPixel == 8 || BitsPerPixel == 4) && interactive_bmp)
+  if ((BitsPerPixel == 8 || BitsPerPixel == 4))
     {
-      if (! save_dialog ())
-        return GIMP_PDB_CANCEL;
+      if (interactive_bmp)
+        {
+          if (! save_dialog ())
+            return GIMP_PDB_CANCEL;
+        }
+      else
+        {
+          encoded = rle_encoded;
+        }
     }
 
   /* Let's take some file */
Only in plug-ins/bmp.orig: .bmpwrite.c.swp
