--- gimp-2.3.9.orig/gimptool-2.0.in
+++ gimp-2.3.9/gimptool-2.0.in
@@ -274,6 +278,7 @@
               test $donothing = "yes" || $cmd
               dest=`echo $1 | sed -e 's#.*/\([^/].*\)$#\1#'`
               cmd="$install_cmd $1 $DESTDIR$install_dir/$dest"
+              mkdir -p $DESTDIR$install_dir
               test $quiet = "yes" || echo $cmd
               test $donothing = "yes" || exec $cmd
             else
@@ -326,6 +331,7 @@
             test $donothing = "yes" || $cmd
           fi
           dest=`echo $1 | sed -e 's#.*/\([^/].*\)$#\1#' -e 's/\.[^.]*$//'`
+          mkdir -p $install_dir
           cmd="$cc $cflags $gimp_cflags -o $install_dir/$dest $1 $ldflags $gimp_libs $libs"
           test $quiet = "yes" || echo $cmd
           test $donothing = "yes" || exec $cmd
