Index: gimp-2.6.7/etc/gimprc
===================================================================
--- gimp-2.6.7.orig/etc/gimprc
+++ gimp-2.6.7/etc/gimprc
@@ -625,7 +625,7 @@
 # appended to the command with a space separating the two.  This is a single
 # filename.
 # 
-# (web-browser "firefox %s")
+(web-browser "sensible-browser %s")
 
 # When enabled, the online user manual will be used by the help system.
 # Otherwise the locally installed copy is used.  Possible values are yes and
