Index: app/widgets/gimphelp.c
===================================================================
--- app/widgets/gimphelp.c	(revision 27111)
+++ app/widgets/gimphelp.c	(revision 27113)
@@ -503,13 +503,13 @@
   *domain_names = g_new0 (gchar *, n_domains + 1);
   *domain_uris  = g_new0 (gchar *, n_domains + 1);
 
-  *domain_names[0] = g_strdup ("http://www.gimp.org/help");
-  *domain_uris[0]  = gimp_help_get_default_domain_uri (gimp);
+  (*domain_names)[0] = g_strdup ("http://www.gimp.org/help");
+  (*domain_uris)[0]  = gimp_help_get_default_domain_uri (gimp);
 
   for (i = 0; i < n_domains; i++)
     {
-      *domain_names[i + 1] = plug_in_domains[i];
-      *domain_uris[i + 1]  = plug_in_uris[i];
+      (*domain_names)[i + 1] = plug_in_domains[i];
+      (*domain_uris)[i + 1]  = plug_in_uris[i];
     }
 
   g_free (plug_in_domains);
Index: app/plug-in/gimppluginprocframe.c
===================================================================
--- app/plug-in/gimppluginprocframe.c	(revision 27111)
+++ app/plug-in/gimppluginprocframe.c	(revision 27113)
@@ -75,7 +75,7 @@
 
   proc_frame->main_context       = g_object_ref (context);
   proc_frame->context_stack      = NULL;
-  proc_frame->procedure          = GIMP_PROCEDURE (procedure);
+  proc_frame->procedure          = procedure ? g_object_ref (procedure) : NULL;
   proc_frame->main_loop          = NULL;
   proc_frame->return_vals        = NULL;
   proc_frame->progress           = progress ? g_object_ref (progress) : NULL;
@@ -118,6 +118,12 @@
       proc_frame->main_context = NULL;
     }
 
+  if (proc_frame->procedure)
+    {
+      g_object_unref (proc_frame->procedure);
+      proc_frame->procedure = NULL;
+    }
+
   if (proc_frame->return_vals)
     {
       g_value_array_free (proc_frame->return_vals);
