diff -Nur gimp-2.4.0~rc2/plug-ins/help/gimphelpdomain.c gimp-2.4.0~rc2.new/plug-ins/help/gimphelpdomain.c
--- gimp-2.4.0~rc2/plug-ins/help/gimphelpdomain.c	2007-08-31 09:39:09.000000000 -0400
+++ gimp-2.4.0~rc2.new/plug-ins/help/gimphelpdomain.c	2007-09-15 11:51:19.000000000 -0400
@@ -176,7 +176,8 @@
             {
               g_message ("%s\n\n%s",
                          _("The GIMP help files are not found."),
-                         _("Please install the additional help package or use "
+                         _("Please install the gimp-help-en package or a "
+							"gimp-help-XX package for your language, or use "
                            "the online user manual at http://docs.gimp.org/."));
             }
           else
