--- gimp-2.3.14/app/display/gimpcanvas.c.orig	2007-02-15 00:36:25.000000000 -0500
+++ gimp-2.3.14/app/display/gimpcanvas.c	2007-02-15 00:31:52.000000000 -0500
@@ -225,7 +225,7 @@
 
     case GIMP_CANVAS_STYLE_XOR:
       mask |= GDK_GC_FUNCTION | GDK_GC_CAP_STYLE | GDK_GC_JOIN_STYLE;
-      values.function   = GDK_XOR;
+      values.function   = GDK_INVERT;
       values.cap_style  = GDK_CAP_NOT_LAST;
       values.join_style = GDK_JOIN_MITER;
       break;
