--- gimp.orig/app/display/gimpdisplayshell-render.c	2007-08-31 09:38:01.000000000 -0400
+++ gimp/app/display/gimpdisplayshell-render.c	2007-09-14 16:09:13.000000000 -0400
@@ -1062,6 +1061,12 @@
   guint         top_weight;
   guint         middle_weight;
   guint         bottom_weight;
+  
+  guint         source_width;
+  guint         source_height;
+
+  source_width = tile_manager_width (info->src_tiles);
+  source_height = tile_manager_height (info->src_tiles);
 
   /* dispatch to fast path functions on special conditions */
   if ((gimp_zoom_quality & GIMP_DISPLAY_ZOOM_FAST)
@@ -1090,7 +1095,8 @@
   else if (((info->src_y)     & ~(TILE_WIDTH -1)) ==
            ((info->src_y + 1) & ~(TILE_WIDTH -1)) &&
            ((info->src_y)     & ~(TILE_WIDTH -1)) ==
-           ((info->src_y - 1) & ~(TILE_WIDTH -1))
+           ((info->src_y - 1) & ~(TILE_WIDTH -1)) &&
+           (info->src_y + 1 < source_height)
           )
     {
       /* all the tiles needed are in a single row, use a tile iterator
@@ -1274,9 +1280,22 @@
 
         center_weight = footprint_x - left_weight - right_weight;
 
-      box_filter (left_weight, center_weight, right_weight,
-                  top_weight, middle_weight, bottom_weight, foosum,
-                  src, dest, bpp);
+        if (src_x + 1 >= source_width)
+          {
+            src[2]=src[1];
+            src[5]=src[4];
+            src[8]=src[7];
+          }
+        if (info->src_y + 1 >= source_height)
+          {
+            src[6]=src[3];
+            src[7]=src[4];
+            src[8]=src[5];
+          }
+
+        box_filter (left_weight, center_weight, right_weight,
+                    top_weight, middle_weight, bottom_weight, foosum,
+                    src, dest, bpp);
       }
       dest += bpp;
 
@@ -1619,6 +1638,10 @@
   guint         middle_weight;
   guint         bottom_weight;
 
+  guint         source_width;
+  
+  source_width = tile_manager_width (info->src_tiles);
+
   footprint_y = (1.0/info->scaley) * 256;
   footprint_x = (1.0/info->scalex) * 256;
   foosum      = footprint_x * footprint_y;
@@ -1750,6 +1773,12 @@
 
         center_weight = footprint_x - left_weight - right_weight;
 
+       if (src_x + 1 >= source_width)
+        {
+           src[2]=src[1];
+           src[5]=src[4];
+           src[8]=src[7];
+        }
         box_filter (left_weight, center_weight, right_weight,
                     top_weight, middle_weight, bottom_weight, foosum,
                     src, dest, bpp);
